File: 2025-01-January-GFG-POTD/January 07 - Pair with given sum in a sorted array/Solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int countPairs(vector<int>& arr, int target) {
        unordered_map<int, int> freq;
        int res = 0;

        for (int num : arr) {
            res += freq[target - num];
            freq[num]++;
        }

        return res;
    }
};

int main() {
    int t;
    cin >> t; // Input the number of test cases
    while (t--) {
        string inputLine;
        cin.ignore(); // To handle the newline after the number of test cases
        getline(cin, inputLine);

        vector<int> arr;
        stringstream ss(inputLine);
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }

        int target;
        cin >> target;

        Solution obj;
        int res = obj.countPairs(arr, target);
        cout << res << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-01-January-GFG-POTD/January 07 - Pair with given sum in a sorted array/Solution.py
from collections import defaultdict

class Solution:
    def countPairs(self, arr, target):
        freq = defaultdict(int)
        res = 0

        for num in arr:
            res += freq[target - num]
            freq[num] += 1

        return res

if __name__ == "__main__":
    t = int(input())  # Input the number of test cases
    for _ in range(t):
        arr = list(map(int, input().strip().split()))
        target = int(input())

        obj = Solution()
        res = obj.countPairs(arr, target)
        print(res)
        print("~")
s
